tests/TestCase/ORM/MarshallerTest.php

```diff
@@ -3241,6 +3241,105 @@ function ($e, $data) {
         $this->assertEquals(1, $entity->tags[1]->_joinData->modified_by);
     }
 
+    /**
+     * Test Model.afterMarshal event.
+     *
+     * @return void
+     */
+    public function testAfterMarshalEvent()
+    {
+        $data = [
+            'title' => 'original title',
+            'body' => 'original content',
+            'user' => [
+                'name' => 'Robert',
+                'username' => 'rob',
+            ],
+        ];
+
+        $marshall = new Marshaller($this->articles);
+
+        $this->articles->getEventManager()->on(
+            'Model.afterMarshal',
+            function ($e, $entity, $data, $options) {
+                $this->assertInstanceOf('Cake\ORM\Entity', $entity);
+                $this->assertArrayHasKey('validate', $options);
+                $this->assertFalse($options['isMerge']);
+
+                $data['title'] = 'Modified title';
+                $data['user']['username'] = 'robert';
+                $options['associated'] = ['Users'];
+
+                $entity->body = 'Modified body';
+            }
+        );
+
+        $entity = $marshall->one($data);
+
+        $this->assertSame('original title', $entity->title, '$data is immutable');
+        $this->assertSame('Modified body', $entity->body);
+        // both $options and $data are unchangeable
+        $this->assertIsArray($entity->user, '$options[\'associated\'] is ignored');
+        $this->assertSame('Robert', $entity->user['name']);
```

Comment:
$entity->user->name right?

Reply:
no 
since `$data` and `$options` are unchangeable
```$options['associated'] = ['Users'];``` will not affect 
so `user` is not a entity